scripts/update-message-index.py

```diff
@@ -0,0 +1,175 @@
+#!/usr/bin/env python3
+
+import argparse
+from itertools import chain
+from pathlib import Path
+from string import Template
+from typing import NamedTuple, Sequence
+import yaml
+import jinja2
+
+
+class Version(NamedTuple):
+    major: int
+    minor: int
+    patch: int = 0
+
+    def __str__(self):
+        return ".".join(map(str, self))
+
+
+class Example(NamedTuple):
+    name: str
+    content: str
+
+
+class Entry(NamedTuple):
+    code: int
+    id: str
+    type: str
+    description: str
+    details: str
+    added: Version
+    removed: Version | None
+    examples: tuple[Example, ...]
+
+
+def parse_version(raw_version: str) -> Version:
+    version = raw_version.split(".")
+    assert 2 <= len(version) <= 3 and all(n.isdecimal() for n in version), raw_version
+    return Version(*map(int, version))
+
+
+VALID_TYPES = ("warning", "error")
+DEFAULT_VERSION = "1.0"  # [TODO] some initial version?
+
+
+def parse_example(entry) -> Example:
+    name = entry.get("name")
+    assert name, entry
+    content = entry.get("content")
+    assert content
+    return Example(name=name, content=content)
+
+
+def parse_entry(entry) -> Entry:
+    code = entry.get("code")
+    assert code is not None, entry
+    id = entry.get("id")
+    assert id is not None, entry
+    type_ = entry.get("type")
+    assert type_ in VALID_TYPES, entry
+    description = entry.get("description")
+    assert description is not None, entry
+    details = entry.get("details", "")
+    added = parse_version(entry.get("added", DEFAULT_VERSION))
+    if removed := entry.get("removed"):
+        removed = parse_version(removed)
+    if examples := entry.get("examples", ()):
+        examples = tuple(map(parse_example, examples))
+    return Entry(
+        code=code,
+        id=id,
+        type=type_,
+        description=description,
+        added=added,
+        removed=removed,
+        details=details,
+        examples=examples,
+    )
+
+
+def skip_removed(entries: Sequence[Entry]):
+    return filter(lambda e: e.removed is None, entries)
+
+
+def make_message_code(entry: Entry) -> str:
+    return f"XKB-{entry.code:0>3}"
```

Comment:
```python
class Entry(NamedTuple):
...
    @property
    def message_code(self) -> str:
        return f"XKB-{entry.code:0>3}"
```
and the same for message_index and name

Reply:
Good idea, although I ended to use them as normal function because they are use as Jinja filters.